### PR TITLE
server: instruct user to set the port in error

### PR DESCRIPTION
**Pull request by justinj**
During a user interview today, a user had something listening on port
8080 when starting Cockroach, and it subsequently failed with an error,
(reasonably) their first instinct was to try --port, but they
actually needed --http-port, because it was the admin ui that had
failed.

I'm not super confident in this fix, I think ideally we'd print out
something a bit friendlier - like "CockroachDB starts up an
administrative UI which needs to listen on a port, but port 8080 was
taken, use --http-port to choose a different port to run it on", but
it's not really clear to me how to safely verify that the error we got
was actually caused by a taken port (we could inspect the error message...),
open to suggestions for how to make
this better, but here I've just gone with a vague addendum that won't
look *too* out of place in an unrelated error.
The Listen function is here, for the curious:
https://github.com/golang/go/blob/da1b83067dfe554ba1c3de564bc4e6a1f1a9bdb0/src/net/dial.go#L554

### CODE REVIEW

File: pkg/server/server.go [hunk 1]
```diff
@@ -481,7 +481,7 @@ func (s *Server) Start(ctx context.Context) error {
 
 	ln, err := net.Listen("tcp", s.cfg.Addr)
 	if err != nil {
-		return err
+		return errors.Errorf("%s, use --port <port> to set the port", err)
```
Comment (a-robinson) [inline pkg/server/server.go]: This is what `errors.Wrap` and `errors.Wrapf` are for, as in `errors.Wrap(err, "use --port <port> to set the port")`.

**Review by a-robinson — COMMENTED** [review]
This seems like it may be misleading if the error isn't a port misuse error (for example, if they specify an invalid `--host` or `--http-host`).

You could either try to:
1. Do error text matching on "address already in use" errors (which is usually somewhat frowned upon but realistically not too risky for OS-level error messages)
1. Just word your error message more carefully such that it mentions which two flags are controlling the address that couldn't be listend on -- i.e. `--host` and `--port` for the first one, and `--http-host` and `--http-port` for the latter.

**Comment by justinj** [comment]
All in all my preference is to go with the error text matching I guess, though it feels a bit nasty, I think the payoff is worthwhile and the fallout if it breaks is negligible. I'll have to do some digging to make sure whatever matching I do works across OSs, I guess.

File: pkg/server/server.go (hunk 1 above)
Comment (justinj) [inline pkg/server/server.go]: Oh nice 👍

**Comment by a-robinson** [comment]
It looks like all the OSes we care about use the same error message, judging by https://github.com/golang/go/search?utf8=%E2%9C%93&q=%22address+already+in+use%22&type=

**Comment by justinj** [comment]
Ok, updated!

**Review by a-robinson — APPROVED** [review]
I might replace "to set the port" with "to use a different port" in the error messages, but LGTM.

Also, I'd wait on @tamird to chime in before merging though since he tends to feel a little more strongly about issues like error text matching.

**Comment by tamird** [comment]
Thanks for holding off.

Indeed, string error matching feels like a bad idea here. What does the error string look like on Windows?

Also, it's fundamentally the case that the server doesn't know anything about CLI flags, so it's weird to embed knowledge of the particular flag into the server package.

It would certainly be better to pass listeners into the server, rather than what we do now, but that's obviously a rather complicated bit of surgery. As a compromise, you could introduce a new exported error type to the server package which would include the "name" (enum?) of the port which listening on produced the error; the CLI could look for that error and then produce a reasonable error string, perhaps using language like "consider changing the <name> port" which would also avoid the need to assert on any particular error text.

---

Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16282#-:-Klebw-NRSwkKV5DrKzx:bk3zgmw)*

**Comment by justinj** [comment]
Ok, updated! PTAL.

**Comment by benesch** [comment]
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/cli/start.go, line 363 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16282#-Klskrvp7NdFtQmaZTZz:-Klskrvp7NdFtQmaZT_-:bolbasw) ([raw file](https://github.com/cockroachdb/cockroach/blob/d934a4905c06042d9df808b7daa4199bf50722e5/pkg/cli/start.go#L363)):*
> ```Go
> 				if le, ok := err.(server.ListenError); ok {
> 					const errorPrefix = "consider changing the port via --"
> 					if le.Port == server.MainPort {
> ```

```go
if le.Addr == serverCfg.Addr {
    // ...
} else if le.Addr == serverCfg.HTTPAddr {
    // ...
}
```

(see next comment)

---

*[pkg/server/server.go, line 451 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16282#-KlslDFdhQ-WZ9BnCp9O:-KlslDFdhQ-WZ9BnCp9P:b-plfbt9) ([raw file](https://github.com/cockroachdb/cockroach/blob/d934a4905c06042d9df808b7daa4199bf50722e5/pkg/server/server.go#L451)):*
> ```Go
> 
> // ListenError is returned from Start when we fail to start listening on either
> // the main Cockroach port or the admin ui port, so that the CLI can instruct
> ```

Also, I think you're better off calling it the "HTTP port", since it serves more than just the admin UI. Debugging pages, profiling pages, etc.

---

*[pkg/server/server.go, line 455 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16282#-KlskaBt57c7K0sbkw4B:-KlskaBuVbsI_-V2TOGP:b4xkf0f) ([raw file](https://github.com/cockroachdb/cockroach/blob/d934a4905c06042d9df808b7daa4199bf50722e5/pkg/server/server.go#L455)):*
> ```Go
> type ListenError struct {
> 	err  error
> 	Port Port
> ```

If you make this `Addr string` instead of `Port`, I think you can avoid introducing the `Port` type. See comment in `start.go`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16282)*

**Comment by justinj** [comment]
Review status: 0 of 2 files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/server/server.go, line 451 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16282#-KlslDFdhQ-WZ9BnCp9O:-KlssYnwgPSD5yCZsZjs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d934a4905c06042d9df808b7daa4199bf50722e5/pkg/server/server.go#L451)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Also, I think you're better off calling it the "HTTP port", since it serves more than just the admin UI. Debugging pages, profiling pages, etc.
</blockquote></details>

Done.

---

*[pkg/server/server.go, line 455 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/16282#-KlskaBt57c7K0sbkw4B:-KlssZ5zcn5K_l4ndrJA:bei2e4x) ([raw file](https://github.com/cockroachdb/cockroach/blob/d934a4905c06042d9df808b7daa4199bf50722e5/pkg/server/server.go#L455)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

If you make this `Addr string` instead of `Port`, I think you can avoid introducing the `Port` type. See comment in `start.go`.
</blockquote></details>

:thumbsup: much better!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16282)*

**Comment by benesch** [comment]
Deferring to @tamird, but LGTM FWIW!

---

Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16282#-:-KlsvZbfe0xbCDnqiJhS:b3a0b9e)*

**Comment by tamird** [comment]
:lgtm:

Have you tested what the new error messages look like? It'd be good to include in the commit message.

---

Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/server/server.go, line 445 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/16282#-KlswyReMAGBuPFWiaIB:-KlswyReMAGBuPFWiaIC:bo2cpoo) ([raw file](https://github.com/cockroachdb/cockroach/blob/fa6fa3e2210e65a64394f76e284481758f8d2b1b/pkg/server/server.go#L445)):*
> ```Go
> // user on what might have gone wrong.
> type ListenError struct {
> 	err  error
> ```

nit: you could embed this, I think

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16282#-:-KlsxBH1Tx2ycOK0A05y:baspyva)*

**Comment by benesch** [comment]
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/16282)*

**Pull request merged**